src/inject/dynamic-theme/image.ts

```diff
@@ -15,22 +16,28 @@ export interface ImageDetails {
     isLarge: boolean;
 }
 
+const imageManager = new AsyncQueue();
+
 export async function getImageDetails(url: string) {
-    let dataURL: string;
-    if (url.startsWith('data:')) {
-        dataURL = url;
-    } else {
-        dataURL = await getImageDataURL(url);
-    }
-    const image = await urlToImage(dataURL);
-    const info = analyzeImage(image);
-    return {
-        src: url,
-        dataURL,
-        width: image.naturalWidth,
-        height: image.naturalHeight,
-        ...info,
-    };
+    return new Promise<ImageDetails>(async (resolve) => {
+        let dataURL: string;
+        if (url.startsWith('data:')) {
+            dataURL = url;
+        } else {
+            dataURL = await getImageDataURL(url);
```

Comment:
Worth measuring what actually causes the problem: `getImageDataURL`, `urlToImage` or `analyzeImage`. Maybe they should be put into a queue too.

One problem with a queue is it can cause out of memory issues: all the images, dataURLs etc will be stored in memory until their turn comes and they can be garbage collected. The solution is... no solution, just balance between memory and synchronous intervals lengths. So worth checking if it causes anything for those problematic websites.

Reply:
`analyzeImage` is problematic due to `canvas 2d`, which is using `drawImage`, however upon drawing it will need to decode an base64, png. All while blocking the execution of any other code. And I've seen cases taking it longer than 50ms to do so. I would prefer if we had some API to draw images which doesn't rely on such function that would block the whole execution. But that's only somewhat available for webgl today.

getImageDataURL has a potential bottleneck to due sending message to the background-script which is actually not that performant somehow

urlToImage seems to have no problems so-far. Other than `.href =` being slow sometimes which we cannot do anything about.